File: Optional(lab5)/src/main/java/com/epam/poker/java/PokerService.java
package com.epam.poker.java;

import java.util.Comparator;
import java.util.List;
import java.util.Optional;

public class PokerService {
    public static Optional<Bid> getWinner(List<Optional<Bid>> playerBids, List<Optional<Bid>> observersBids) {

        return playerBids.stream().
                flatMap(Optional::stream).
                max(Comparator.comparingInt(Bid::getAmount))
                .or(() ->
                        observersBids.stream()
                                .flatMap(Optional::stream)
                                .max(Comparator.comparingInt(Bid::getAmount))
                );

    }
}

File: Optional(lab5)/src/main/java/com/epam/poker/scala/PokerService.scala
package com.epam.poker.scala

import scala.util.Try

object PokerService {
  def getWinner(playersBids:List[Option[Bid]], observersBids:List[Option[Bid]]) = {
    Try(
      playersBids
        .flatten
        .maxBy(_.amoutn)
    )
      .getOrElse(
        Try(
          observersBids
            .flatten
            .maxBy(_.amoutn)
        ).
          getOrElse("The is no winners")
      )
  }
}
